src/apple/macos/disk.rs

```diff
@@ -1,200 +1,125 @@
+            ffi::kIOMasterPortDefault,
+            matching.cast(),
+            &mut service_iterator,
+        )
+    } != libc::KERN_SUCCESS
+    {
+        return None;
     }
-}
 
-unsafe fn get_str_value(dict: CFDictionaryRef, key: &[u8]) -> Option<String> {
-    get_dict_value(dict, key, |v| {
-        let v = v as cfs::CFStringRef;
-
-        let len_utf16 = cfs::CFStringGetLength(v);
-        let len_bytes = len_utf16 as usize * 2; // Two bytes per UTF-16 codepoint.
-
-        let v_ptr = cfs::CFStringGetCStringPtr(v, cfs::kCFStringEncodingUTF8);
-        if v_ptr.is_null() {
-            // Fallback on CFStringGetString to read the underlying bytes from the CFString.
-            let mut buf = vec![0; len_bytes];
-            let success = cfs::CFStringGetCString(
-                v,
-                buf.as_mut_ptr(),
-                len_bytes as _,
-                cfs::kCFStringEncodingUTF8,
-            );
-
-            if success != 0 {
-                utils::vec_to_rust(buf)
-            } else {
-                None
-            }
-        } else {
-            utils::cstr_to_rust_with_size(v_ptr, Some(len_bytes))
-        }
-    })
-}
+    // Safety: We checked for success, so there is always a valid iterator, even if its empty.
+    let service_iterator = unsafe { IOReleaser::new_unchecked(service_iterator) };
 
-unsafe fn get_bool_value(dict: CFDictionaryRef, key: &[u8]) -> Option<bool> {
-    get_dict_value(dict, key, |v| Some(v as CFBooleanRef == kCFBooleanTrue))
-}
+    let mut parent_entry: ffi::io_registry_entry_t = 0;
 
-fn new_disk(
-    name: OsString,
-    mount_point: PathBuf,
-    type_: DiskType,
-    is_removable: bool,
-) -> Option<Disk> {
-    let mount_point_cpath = to_cpath(&mount_point);
-    let mut total_space = 0;
-    let mut available_space = 0;
-    let mut file_system = None;
-    unsafe {
-        let mut stat: statfs = mem::zeroed();
-        if statfs(mount_point_cpath.as_ptr() as *const i8, &mut stat) == 0 {
-            // APFS is "special" because its a snapshot-based filesystem, and modern
-            // macOS devices take full advantage of this.
-            //
-            // By default, listing volumes with `statfs` can return both the root-level
-            // "data" partition and any snapshots that exist. However, other than some flags and
-            // reserved(undocumented) bytes, there is no difference between the OS boot snapshot
-            // and the "data" partition.
-            //
-            // To avoid duplicating the number of disks (and therefore available space, etc), only return
-            // a disk (which is really a partition with APFS) if it is the root of the filesystem.
-            let is_root = stat.f_flags & libc::MNT_ROOTFS as u32 == 0;
-            if !is_root {
-                return None;
+    while let Some(mut current_service_entry) =
+        IOReleaser::new(unsafe { ffi::IOIteratorNext(service_iterator.inner()) })
+    {
+        loop {
```

Comment:
It'll always either `break` (to continue le `while let`) or `return`. I think this `loop` can be removed then, no?

Reply:
This loop still needs to exist AFAIK. Without this loop in place, we don't recurse far enough through the IO registry entry's properties to find the disk information we're looking for. For example if I remove this, the function starts becoming unable to get the correct disk information since after a single call to `IORegistryEntryCreateCFProperty` returns nothing, it quits.

Reply:
Would it help if I added a comment to this effect?

Reply:
I think I got it but a comment would indeed be very welcome. :+1:

Reply:
I've added a comment about this to the loop so that those looking at the code in the future have more context 👍.